### PR TITLE
Add Requeued Mode with sheduled Time

### PR DESCRIPTION
**Pull request by Houtmann**
In response to pull request #278

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -524,6 +524,20 @@ Defaults to ``None``. This option is useful if you want to redirect all emails t
...
+Mail Retry
+-------------------
+
+Defaults to Not Activated. Auto requeue failed email with time interval
+
+.. code-block:: python
+
+    # Put this in settings.py
+    POST_OFFICE = {
+     'MAIL_MAX_RETRY': 4
```
Comment (selwin) [inline README.rst:536]: Mind changing this to `MAX_RETRIES`?

File: post_office/mail.py [hunk 2]
```diff
@@ -270,7 +272,26 @@ def send(email):
     Email.objects.filter(id__in=email_ids).update(status=STATUS.sent)
 
     email_ids = [email.id for (email, e) in failed_emails]
-    Email.objects.filter(id__in=email_ids).update(status=STATUS.failed)
+
+    numbers_mails_get_back_in_queue = 0
+    # Retry is enabled
+    if get_max_retry() > 0:
```
Comment (selwin) [inline post_office/mail.py:278]: `if get_max_retries():` to make it more pythonic.

File: post_office/mail.py [hunk 3]
```diff
@@ -270,7 +272,26 @@ def send(email):
...
+        emails_failed = Email.objects.filter(id__in=email_ids)
+        for email in emails_failed:
+            if email.number_of_retries < get_max_retry():
```
Comment (selwin) [inline post_office/mail.py:281]: We should cache this in a variable `max_retries = get_max_retries()` so we don't keep calling this function in a loop.

File: post_office/mail.py [hunk 4]
```diff
@@ -270,7 +272,26 @@ def send(email):
...
+                email.scheduled_time = now() + get_time_delta_to_retry()
+                if email.number_of_retries is None:
+                    email.number_of_retries = 1
+                else:
+                    email.number_of_retries = email.number_of_retries + 1
+                numbers_mails_get_back_in_queue += 1
+            else:
+                email.status = STATUS.failed
+
+            email.save()
```
Comment (selwin) [inline post_office/mail.py:292]: We should use `bulk_update()` for this. https://docs.djangoproject.com/en/3.0/ref/models/querysets/#bulk-update

File: post_office/mail.py [hunk 5]
```diff
@@ -270,7 +272,26 @@ def send(email):
     Email.objects.filter(id__in=email_ids).update(status=STATUS.sent)
 
     email_ids = [email.id for (email, e) in failed_emails]
-    Email.objects.filter(id__in=email_ids).update(status=STATUS.failed)
+
+    numbers_mails_get_back_in_queue = 0
```
Comment (selwin) [inline post_office/mail.py:276]: Can we change this to `requeued_emails`?

**Comment by Houtmann** [comment]
Yeah, I'm doing it this week.
Thanks for this review

File: README.rst (hunk 1 above)
Comment (Houtmann) [reply README.rst]: ok

File: post_office/mail.py (hunk 2 above)
Comment (Houtmann) [reply post_office/mail.py]: ok

File: post_office/mail.py (hunk 3 above)
Comment (Houtmann) [reply post_office/mail.py]: ok

File: post_office/mail.py (hunk 4 above)
Comment (Houtmann) [reply post_office/mail.py]: ok
Same thing for requeued email

File: post_office/mail.py (hunk 5 above)
Comment (Houtmann) [reply post_office/mail.py]: ok

**Comment by Houtmann** [comment]
I think my last commit it's ok ;)

File: post_office/mail.py [hunk 6]
```diff
@@ -270,7 +272,35 @@ def send(email):
...
+        emails_failed = Email.objects.filter(id__in=email_ids)
+
+        emails_max_retries_exceeded = []  # emails status will change to failed
+        emails_to_requeue = []  # emails status will change to requeued
+
+        for email in emails_failed:
+            if email.number_of_retries < max_retries:
+                email.status = STATUS.requeued
+                email.scheduled_time = now() + get_time_delta_to_retry()
+                if email.number_of_retries is None:
```
Comment (selwin) [inline post_office/mail.py:289]: We shouldn't need to check for `None` since the `number_of_retries` column is not nullable.

Comment (selwin) [reply post_office/mail.py:289]: So we either change this, or make the `number_of_retries` column nullable, which may speed up schema migrations on dataabases and also consume less space (I think).

File: post_office/tests/test_mail.py [hunk 7]
```diff
@@ -395,4 +396,32 @@ def test_send_bulk_with_faulty_template(self):
...
+        previous_settings = settings.POST_OFFICE
+        setattr(settings, 'POST_OFFICE', {'MAX_RETRIES': 2})
+
+        self.assertEqual(get_max_retries(), 2)
+        self.assertEqual(get_time_delta_to_retry(), timedelta(minutes=15))
+
+
+        email = Email.objects.create(to='to@example.com', from_email='from@example.com', status=STATUS.queued, backend_alias='error')
+        _send_bulk([email], uses_multiprocessing=False)
+        email = Email.objects.get(id=email.id)
```
Comment (selwin) [inline post_office/tests/test_mail.py:411]: Nitpick: we can do `email.refresh_from_db()` here.

File: post_office/tests/test_mail.py [hunk 8]
```diff
@@ -395,4 +396,32 @@ def test_send_bulk_with_faulty_template(self):
...
+        email.save()
+        _send_bulk([email], uses_multiprocessing=False)
+        self.assertEqual(email.status, STATUS.requeued)
+
+        email.scheduled_time = now()
+        email.save()
+
+        _send_bulk([email], uses_multiprocessing=False)
+        self.assertEqual(email.status, STATUS.requeued)
+
```
Comment (selwin) [inline post_office/tests/test_mail.py:426]: Can you add another call to `_send_bulk()`? After this call, `email.status` should be set to `failed`

File: post_office/mail.py (hunk 6 above)
Comment (Houtmann) [reply post_office/mail.py]: I put in the null column and remove defaut, but I had to refactor the code a little bit.

File: post_office/tests/test_mail.py (hunk 7 above)
Comment (Houtmann) [reply post_office/tests/test_mail.py]: ;)

File: post_office/tests/test_mail.py (hunk 8 above)
Comment (Houtmann) [reply post_office/tests/test_mail.py:426]: It's ok :D

File: post_office/tests/test_mail.py [hunk 9]
```diff
@@ -398,6 +399,37 @@ def test_send_bulk_with_faulty_template(self):
...
+        self.assertEqual(get_max_retries(), 2)
+        self.assertEqual(get_time_delta_to_retry(), timedelta(minutes=15))
+
+
+        email = Email.objects.create(to='to@example.com', from_email='from@example.com', status=STATUS.queued, backend_alias='error')
+        _send_bulk([email], uses_multiprocessing=False)
+        email.refresh_from_db()
+
+        self.assertEqual(int(email.scheduled_time.timestamp()), int(now().__add__(timedelta(minutes=15)).timestamp()))
```
Comment (jrief) [inline post_office/tests/test_mail.py:414]: By relying on `now()`, you add unpredictable behaviour, and it becomes impossible to step through the testing code using a debugger.

**Comment by jrief** [comment]
@Houtmann Thanks for this pull request, this is a really important feature! I used it as starting point for #315 which fixes some issues I found, while reviewing it. Therefore I will close it in favor of the latter.

**Comment by Houtmann** [comment]
OK it's good.

**Pull request closed without merging**